src/workerd/api/node/zlib-util.h

```diff
@@ -250,38 +291,113 @@ class ZlibUtil final: public jsg::Object {
         kj::Array<kj::byte> output,
         int outputOffset,
         int outputLength);
-    void writeSync(jsg::Lock& js,
-        int flush,
-        jsg::Optional<kj::Array<kj::byte>> input,
-        int inputOffset,
-        int inputLength,
-        kj::Array<kj::byte> output,
-        int outputOffset,
-        int outputLength);
-    void params(jsg::Lock& js, int level, int strategy);
     void reset(jsg::Lock& js);
 
+    JSG_RESOURCE_TYPE(CompressionStream) {
+      JSG_METHOD(close);
+      JSG_METHOD_NAMED(write, template write<true>);
+      JSG_METHOD_NAMED(writeSync, template write<false>);
+      JSG_METHOD(reset);
+      JSG_METHOD(setErrorHandler);
+    }
+
+  protected:
+    CompressionContext* context() {
+      return &context_;
+    }
+
+    void initializeStream(jsg::BufferSource _write_result, jsg::Function<void()> writeCallback);
+
+    // Allocation functions provided to zlib itself. We store the real size of
+    // the allocated memory chunk just before the "payload" memory we return
+    // to zlib.
+    static void* AllocForZlib(void* data, uInt items, uInt size);
+    static void* AllocForBrotli(void* data, size_t size);
+    static void FreeForZlib(void* data, void* pointer);
+
+  private:
+    // Used to store allocations in Brotli* operations.
+    // This declaration should be physically positioned before
+    // context to avoid `heap-use-after-free` ASan error.
```

Comment:
non-blocking nit: I would expand the comment slight to connect this more concretely with the `AllocFor_` and `FreeFor_` APIs above.